src/gprs.c

```diff
@@ -536,9 +538,171 @@ static void pri_context_signal_settings(struct pri_context *ctx,
 				context_settings_append_ipv6);
 }
 
-static void pri_parse_proxy(struct pri_context *ctx, const char *proxy)
+static void set_default_dns(char **dns_servers)
 {
-	char *scheme, *host, *port, *path;
+	struct in_addr server_in_addr;
+	int i;
+
+	/*
+	 * In this function we modify the global variable _res from resolv.h to
+	 * be able to use a specific DNS server to resolve the MMS proxy/MMSC
+	 * address. This is a hack, although used in some places. Also, the IPv6
+	 * case cannot be handled this way. ldns is a possible option
+	 * (http://www.nlnetlabs.nl/projects/ldns/), but it would introduce a
+	 * new dependence. TODO: search alternatives.
+	 */
+	_res.nscount = 0;
+
+	for (i = 0; dns_servers[i] != NULL; ++i) {
+		if (inet_pton(AF_INET, dns_servers[i], &server_in_addr) != 1)
+			continue;
+
+		_res.nsaddr_list[_res.nscount].sin_addr = server_in_addr;
+		++(_res.nscount);
+
+		/* Size of nsaddr_list array (see resolv.h) */
+		if (_res.nscount == MAXNS)
+			break;
+	}
+}
+
+static gboolean lookup_address(const struct context_settings *settings,
+						const char *proxy,
+						struct sockaddr_in *addr)
+{
+	static gboolean first_time = TRUE;
+	struct addrinfo *result = NULL;
+	struct addrinfo hint;
+	int rc, ret = FALSE;
+
+	DBG("hostname is %s", proxy);
+
+	/*
+	 * We need to initialize the resolv library before changing the DNS
+	 * server. gethostbyname("localhost") does that and at the same time
+	 * does not generate network traffic. Calling directly res_init does
+	 * not really work: that function is called internally again in the
+	 * first call to a resolv function.
+	 */
+	if (first_time) {
+		gethostbyname("localhost");
+		first_time = FALSE;
+	}
+
+	set_default_dns(settings->ipv4->dns);
+
+	memset(&hint, 0, sizeof(hint));
+	hint.ai_socktype = SOCK_STREAM;
+	rc = getaddrinfo(proxy, NULL /* service */, &hint, &result);
```

Comment:
One of the other issues with this approach is that getaddrinfo is a blocking function, and on a slow mobile data connection ( think 2g ), a DNS lookup could take several seconds ( or more ).  I also couldn't find a defined timeout for getaddrinfo, it's possible that this call could hang indefinitely.  If there is a default timeout, it'd be good to mention it in a code comment.

Finally, there is an asynchronous version of this function called getaddrinfo_a, however I'm not sure how compatible it is with the glib's mainloop.   It offers the ability to notify the caller of a resolved name via a signal or via a callback in a new thread.  It's probably worth some investigation.